transport/internet/domainsocket/listener.go

```diff
@@ -0,0 +1,214 @@
+package domainsocket
+
+import (
+	"context"
+	"net"
+	"os"
+	"syscall"
+	"time"
+
+	"v2ray.com/core/common/bitmask"
+)
+
+type Listener struct {
+	ln           net.Listener
+	listenerChan chan<- net.Conn
+	ctx          context.Context
+	path         string
+	lockfile     *os.File
+	state        bitmask.Byte
+	cancal       func()
+}
+
+const (
+	STATE_UNDEFINED   = 0
+	STATE_INITIALIZED = 1 << iota
+	STATE_LOWERUP     = 1 << iota
+	STATE_UP          = 1 << iota
+	STATE_TAINT       = 1 << iota
+)
+
+func ListenDS(ctx context.Context, path string) (*Listener, error) {
+
+	vln := &Listener{path: path, state: STATE_INITIALIZED, ctx: ctx}
+	return vln, nil
+}
+
+func (ls *Listener) Down() error {
+	var err error
+	if !ls.state.Has(STATE_LOWERUP | STATE_UP) {
+		err = newError(ls.state).Base(newError("Invalid State:Down"))
+		if ___DEBUG_PANIC_WHEN_ENCOUNTED_IMPOSSIBLE_ERROR {
+			panic(err)
+		}
+		return err
+	}
+
+	ls.cancal()
+	closeerr := ls.ln.Close()
+	var lockerr error
+	if isUnixDomainSocketFileSystemBased(ls.path) {
+		lockerr = giveupLock(ls.lockfile)
+	}
+	if closeerr != nil && lockerr != nil {
+		if ___DEBUG_PANIC_WHEN_ERROR_UNPROPAGATEABLE {
+			panic(closeerr.Error() + lockerr.Error())
+		}
+	}
+
+	if closeerr != nil {
+		return newError("Cannot Close Unix domain socket listener").Base(closeerr)
+	}
+	if lockerr != nil {
+		return newError("Cannot release lock for Unix domain socket listener").Base(lockerr)
+	}
+	ls.state.Clear(STATE_LOWERUP | STATE_UP)
+	return nil
+}
+
+//LowerUP Setup systen level Listener
+func (ls *Listener) LowerUP() error {
+	var err error
+
+	if !ls.state.Has(STATE_INITIALIZED) || ls.state.Has(STATE_LOWERUP) {
+		err = newError(ls.state).Base(newError("Invalid State:LowerUP"))
+		if ___DEBUG_PANIC_WHEN_ENCOUNTED_IMPOSSIBLE_ERROR {
+			panic(err)
+		}
+		return err
+	}
+
+	if isUnixDomainSocketFileSystemBased(ls.path) && !___DEBUG_IGNORE_FLOCK {
```

Comment:
If the unix domain socket is filesystem based, an file lock must be used to claim the right for listening on respective file.

Reply:
https://gavv.github.io/blog/unix-socket-reuse/

Reply:
Please put this comment into source code.